mos-platform/nes-gtrom/_prg-rom-banked.ld

```diff
@@ -0,0 +1,259 @@
+/* Automatically generated by generate-rom-banked-sections.lua. */
+
+INPUT(reset.o)
+__prg_rom_0_lma = 0x008000;
+__prg_rom_1_lma = 0x018000;
+__prg_rom_2_lma = 0x028000;
+__prg_rom_3_lma = 0x038000;
+__prg_rom_4_lma = 0x048000;
+__prg_rom_5_lma = 0x058000;
+__prg_rom_6_lma = 0x068000;
+__prg_rom_7_lma = 0x078000;
+__prg_rom_8_lma = 0x088000;
+__prg_rom_9_lma = 0x098000;
+__prg_rom_10_lma = 0x0a8000;
+__prg_rom_11_lma = 0x0b8000;
+__prg_rom_12_lma = 0x0c8000;
+__prg_rom_13_lma = 0x0d8000;
+__prg_rom_14_lma = 0x0e8000;
+__prg_rom_15_lma = 0x0f8000;
+
+__prg_rom_0_offset = 0x000000;
+__prg_rom_1_offset = 0x008000;
+__prg_rom_2_offset = 0x010000;
+__prg_rom_3_offset = 0x018000;
+__prg_rom_4_offset = 0x020000;
+__prg_rom_5_offset = 0x028000;
+__prg_rom_6_offset = 0x030000;
+__prg_rom_7_offset = 0x038000;
+__prg_rom_8_offset = 0x040000;
+__prg_rom_9_offset = 0x048000;
+__prg_rom_10_offset = 0x050000;
+__prg_rom_11_offset = 0x058000;
+__prg_rom_12_offset = 0x060000;
+__prg_rom_13_offset = 0x068000;
+__prg_rom_14_offset = 0x070000;
+__prg_rom_15_offset = 0x078000;
+
+MEMORY {
+  /* PRG-ROM LMAs. */
+  prg_rom_0 : ORIGIN = __prg_rom_0_lma, LENGTH = __prg_rom_size >= 32 ? 0x8000 - 0x6 : 0
+  prg_rom_1 : ORIGIN = __prg_rom_1_lma, LENGTH = __prg_rom_size >= 64 ? 0x8000 - 0x6 : 0
+  prg_rom_2 : ORIGIN = __prg_rom_2_lma, LENGTH = __prg_rom_size >= 128 ? 0x8000 - 0x6 : 0
+  prg_rom_3 : ORIGIN = __prg_rom_3_lma, LENGTH = __prg_rom_size >= 128 ? 0x8000 - 0x6 : 0
+  prg_rom_4 : ORIGIN = __prg_rom_4_lma, LENGTH = __prg_rom_size >= 256 ? 0x8000 - 0x6 : 0
+  prg_rom_5 : ORIGIN = __prg_rom_5_lma, LENGTH = __prg_rom_size >= 256 ? 0x8000 - 0x6 : 0
+  prg_rom_6 : ORIGIN = __prg_rom_6_lma, LENGTH = __prg_rom_size >= 256 ? 0x8000 - 0x6 : 0
+  prg_rom_7 : ORIGIN = __prg_rom_7_lma, LENGTH = __prg_rom_size >= 256 ? 0x8000 - 0x6 : 0
+  prg_rom_8 : ORIGIN = __prg_rom_8_lma, LENGTH = __prg_rom_size >= 512 ? 0x8000 - 0x6 : 0
+  prg_rom_9 : ORIGIN = __prg_rom_9_lma, LENGTH = __prg_rom_size >= 512 ? 0x8000 - 0x6 : 0
+  prg_rom_10 : ORIGIN = __prg_rom_10_lma, LENGTH = __prg_rom_size >= 512 ? 0x8000 - 0x6 : 0
+  prg_rom_11 : ORIGIN = __prg_rom_11_lma, LENGTH = __prg_rom_size >= 512 ? 0x8000 - 0x6 : 0
+  prg_rom_12 : ORIGIN = __prg_rom_12_lma, LENGTH = __prg_rom_size >= 512 ? 0x8000 - 0x6 : 0
+  prg_rom_13 : ORIGIN = __prg_rom_13_lma, LENGTH = __prg_rom_size >= 512 ? 0x8000 - 0x6 : 0
+  prg_rom_14 : ORIGIN = __prg_rom_14_lma, LENGTH = __prg_rom_size >= 512 ? 0x8000 - 0x6 : 0
+  prg_rom_15 : ORIGIN = __prg_rom_15_lma, LENGTH = __prg_rom_size >= 512 ? 0x8000 - 0x6 : 0
+  vectors : ORIGIN = 0x10000 - 0x6, LENGTH = 6
+
+  /* CHR-ROM LMAs. */
+  chr_rom_0   : ORIGIN = 0x01000000, LENGTH = __chr_rom_size >= 8    ? 0x2000 : 0
+  chr_rom_1   : ORIGIN = 0x01002000, LENGTH = __chr_rom_size >= 16   ? 0x2000 : 0
+}
+
+REGION_ALIAS("c_readonly", prg_rom_0)
```

Comment:
A silent invariant that we've maintained is that default C sections code must go into a fixed region. The C runtime broadly assumes that libcalls are equally accessible from any function, and libcalls will intrinsically always be placed into the base C sections. More broadly, the C runtime model basically assumes flat addresses; so the C enclave in a environment where that isn't true should generally be the maximal subset of that environment that can be made to have flat addresses.

A way of using C that doesn't invoke libcalls isn't formally specified, and we probably shouldn't start. The implications of all this would be that there is no guaranteed safe way of writing C in contexts where the libcalls might be banked out.

A far code model would help with this; the C flat addressing model can be restored if addresses contain the information necessarily to silently bank libcalls in and out.

Ostensibly this is true for RAM used by libcalls too (although I don't know whether there actually is any at present; I've tried to keep that as minimal as possible); placing the default C writeable sections in PRG-RAM can be just as dangerous. This is always an explicit action on the part of the user though, and unbanked PRG-RAM poses no obstacle for this.

Another thing we could do is carefully place libcalls in sections naming them as such; then mappers like this one could keep them fixed while allowing the rest to vary. Up to you whether it's worth it for this one; we could go back and do this later. Even with a far code model, that might still be a good idea; it would allow summarily marking libcalls near, saving banking overhead for common operations.

Reply:
I apologize, this should have been more documented: Everything that is placed in the `prg_rom_0` region *before*  the `.prg_rom_fixed : {` section is declared (yes, in file parsing order) will become part of the faux "fixed" bank. This works correctly, as all the C sections are allocated to the `c_readonly` region; the PRG-ROM swapping tests rely on this behaviour.

The invariant is correct to maintain, I just forgot to document it.

Reply:
Added a comment which provides the necessary clarification.

Reply:
Sorry for diving right into a whole spiel; I should have probably asked a question here. 😅